public/app/features/datasources/components/picker/DataSourceModal.tsx

```diff
@@ -73,6 +75,42 @@ export function DataSourceModal({
     [analyticsInteractionSrc]
   );
 
+  const grafanaDS = useDatasource('-- Grafana --');
+
+  const onFileDrop = (acceptedFiles: File[], fileRejections: FileRejection[], event: DropEvent) => {
+    if (!grafanaDS) {
+      return;
+    }
+    DFImport.filesToDataframes(acceptedFiles).subscribe(async (next) => {
+      const snapshot: DataFrameJSON[] = [];
+      next.dataFrames.forEach((df: DataFrame) => {
+        const dataframeJson = dataFrameToJSON(df);
+        snapshot.push(dataframeJson);
+      });
+      await onChange(grafanaDS, [
+        {
+          refId: 'A',
+          datasource: {
+            type: 'grafana',
+            uid: 'grafana',
+          },
+          queryType: GrafanaQueryType.Snapshot,
+          snapshot: snapshot,
+          file: next.file,
+        },
```

Comment:
We're repeating the structure of GrafanaQuery in different places. Probably is a good candidate to extract into a function.

Reply:
it also feels like this should be part of the Grafana data source not this modal